internal/app/readycheck.go

```diff
@@ -1,13 +1,135 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package app
 
-import "net/http"
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
 
-func (a *Application) readinessHandler(w http.ResponseWriter, r *http.Request) {
+	"github.com/gardener/etcd-wrapper/internal/bootstrap"
+	"github.com/gardener/etcd-wrapper/internal/types"
+	"github.com/gardener/etcd-wrapper/internal/util"
 
-}
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+)
 
-// SetupReadinessProbe sets up the readiness probe for this application.
+const (
+	// ReadyServerPort is the port number for the server that serves the readiness probe
+	ReadyServerPort       = int64(9095)
+	etcdConnectionTimeout = 5 * time.Second
+	etcdGetTimeout        = 5 * time.Second
+	etcdQueryInterval     = 2 * time.Second
+	etcdEndpointPort      = "2379"
+)
+
+// SetupReadinessProbe sets up the readiness probe for this application. It is a blocking function and therefore
+// the consumer should always call this within a go-routine unless the caller itself wants to block on this which is unlikely.
 func (a *Application) SetupReadinessProbe() {
-	//the caller should call this in a go-routine as this method will setup a HTTP router and call ListenAndServe which blocks.
-	//If the http server errors out then you will have to panic and that should cause the container to exit and then be restarted by kubelet.
+	// Start go routine to regularly query etcd to update its readiness status.
+	go a.queryAndUpdateEtcdReadiness()
+	// If the http server errors out then you will have to panic and that should cause the container to exit and then be restarted by kubelet.
+	if a.isTLSEnabled() {
+		http.Handle("/readyz", http.HandlerFunc(a.readinessHandler))
+		err := http.ListenAndServeTLS(fmt.Sprintf(":%d", ReadyServerPort), a.cfg.ClientTLSInfo.CertFile, a.cfg.ClientTLSInfo.KeyFile, nil)
+		if err != nil {
+			a.logger.Fatal("failed to start TLS readiness endpoint", zap.Error(err))
+		}
+	} else {
+		http.Handle("/readyz", http.HandlerFunc(a.readinessHandler))
+		err := http.ListenAndServe(fmt.Sprintf(":%d", ReadyServerPort), nil)
+		if err != nil {
+			a.logger.Fatal("failed to start readiness endpoint", zap.Error(err))
+		}
+	}
+}
+
+// queryAndUpdateEtcdReadiness periodically queries the etcd DB to check its readiness and updates the status
+// of the query into the etcdStatus struct. It stops querying when the application context is cancelled.
+func (a *Application) queryAndUpdateEtcdReadiness() {
+	// Create a ticker to periodically query etcd readiness
+	ticker := time.NewTicker(etcdQueryInterval)
+	defer ticker.Stop()
+
+	for {
+		// Query etcd readiness and update the status
+		a.etcdReady = a.isEtcdReady()
+		a.logger.Info("etcd ready?", zap.Bool("ready-status", a.etcdReady))
```

Comment:
Please remove this logs as it is making it difficult to check actual etcd cluster members logs
```
{"level":"info","ts":"2023-07-07T11:03:15.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:17.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:19.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:21.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:23.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:25.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:27.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:29.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:31.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:33.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:35.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:37.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:39.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:41.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:43.517Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:45.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:47.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:49.516Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
{"level":"info","ts":"2023-07-07T11:03:51.515Z","caller":"app/readycheck.go:72","msg":"etcd ready?","ready-status":true}
```

Reply:
Thanks for pointing this out
Removed